sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,465 @@
+
+    # Step 5 : Find poles and valuation at infinity
+    poles = find_poles(num, den, x)
+    inf_mul = poles.get(oo, 0)
+    poles, muls = list(poles.keys()), list(poles.values())
+    val_inf = val_at_inf(num, den, x)
+
+    if len(poles):
+        # Check necessary conditions (outlined in the module docstring)
+        if val_inf%2 != 0 or not all(map(lambda mul: (mul == 1 or (mul%2 == 0 and mul >= 2)), muls)):
+            raise ValueError("Rational Solution doesn't exist")
+
+        # Step 6
+        # Construct c-vectors for each singular point
+        c = construct_c(num, den, x, poles, muls)
+
+        # Construct d vectors for each singular point
+        d = construct_d(num, den, a, x, val_inf, inf_mul)
+
+        # Step 7 : Iterate over all possible combinations and return solutions
+        for it in range(2**(len(poles) + 1)):
+            # For each possible combination, generate an array of 0's and 1's
+            # where 0 means pick 1st choice and 1 means pick the second choice.
+            choice = list(map(lambda x: int(x), bin(it)[2:].zfill(len(poles) + 1)))
+
+            # Step 8 and 9 : Compute m and ybar
+            m, ybar, exists = compute_degree(x, poles, choice, c, d, -val_inf//2)
+
+            # Step 10 : Check if a valid solution exists. If yes, also check
+            # if m is a non-negative integer
+            if exists and S(m).is_nonnegative == True and S(m).is_integer == True:
+
+                # Step 11 : Find polynomial solutions of degree m for the auxiliary equation
+                psol, coeffs, exists = solve_aux_eq(a, x, m, ybar)
+
+                # Step 12 : If valid polynomial solution exists, append solution.
+                if exists:
+                    # m == 0 case
+                    if psol == 1 and coeffs == 0:
+                        # p(x) = 1, so p'(x)/p(x) term need not be added
+                        presol.append(ybar)
+                    # m is a positive integer and there are valid coefficients
+                    elif len(coeffs):
+                        # Substitute the valid coefficients to get p(x)
+                        psol = psol.subs(coeffs)
+                        # y(x) = ybar(x) + p'(x)/p(x)
+                        presol.append(ybar + psol.diff(x)/psol)
+
+            # If m is an expression with symbols, we generate a Piecewise solution where
+            # the solution will exist only if the expression is a nonnegative integer.
+            # NOTE: In this case, it seems like x**m + C1 is always a solution
+            # for the auxiliary equation. It is however NOT mentioned in the thesis.
+            elif len(m.free_symbols):
+                C1 = Symbol('C1')
+                psol = x**m + C1
+                presol.append(ybar + psol.diff(x)/psol)
+    # Step 15 : Inverse transform the solutions of the equation in normal form
+    sol = list(map(lambda y: Eq(fx, -y/b2 - b2.diff(x)/(2*b2**2) - b1/(2*b2)), presol))
+    return sol
```

Comment:
A list comprehension is generally easier to read than a `list(map(lambda:` so I would write this as:
```python
sol = [Eq(fx, -y/b2 - b2.diff(x)/(2*b2**2) - b1/(2*b2)) for y in presol]
```
Then again you are recomputing the saem value at each iteration of the loop so:
```python
bp = b2.diff(x)/(2*b2**2) - b1/(2*b2)
sol = [Eq(fx, -y/b2 - bp) for y in presol]
```
I would probably have made a function for these transformations anyway. As a separate function it can be tested separately. You can also test for a large range of (random) inputs that the transformation and inverse transformation do actually invert each other.